### PR TITLE
Split Optin annotation to feature level

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- N/A

## Description
-  Deprecates the project level opt-in annotation `PartiQLExperimental` and split it into feature level. `ExperimentalPartiQLCompilerPipeline` and `ExperimentalWindowFunctions`.
- Add an `experimental()` function to initialize a compiler with the experimental features. (currently just the window operator). 
- Opt-in annotation at line level in example file. 

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  - Yes. 
- Any backward-incompatible changes? **[YES/NO]**
  - Yes. The old project level opt-in annotation is deprecated. 
  - < For this purpose, we define backward-incompatible changes as changes that—when consumed—can potentially result in
errors for users that are using our public APIs or the entities that have `public` visibility in our code-base. >
- Any new external dependencies? **[YES/NO]**
  - No. 
  - < In addition, please also mention any other alternatives you've considered and the reason they've been discarded >

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/main/kotlin/org/partiql/lang/compiler/PartiQLCompilerBuilder.kt [hunk 1]
```diff
@@ -84,12 +87,25 @@ class PartiQLCompilerBuilder private constructor() {
...
             WindowRelationalOperatorFactoryDefault
         )
 
         @JvmStatic
         fun standard() = PartiQLCompilerBuilder()
+
+        @ExperimentalWindowFunctions
+        @JvmStatic
+        // This is a hack to inject the experimental operators to the compiler.
+        fun experimental() = PartiQLCompilerBuilder().customOperatorFactories(EXPERIMENTAL_RELATIONAL_OPERATOR_FACTORIES)
```
Comment (vgapeyev) [inline lang/src/main/kotlin/org/partiql/lang/compiler/PartiQLCompilerBuilder.kt:108]: As we discussed in-person, this method would be addressing some sort of a possible need of a PartQL-*language* customer who might be worried about running code containing features we consider experimental, even if the customer does not use these features. 

This need is different from what @OptIn is for, which is for communicating to a PartiQL-*library* customer that certain of our APIs are not final yet. 

So, let's go without `experimental()` for now.
Instead, WindowRelationalOperatorFactoryDefault should just be added to DEFAULT_RELATIONAL_OPERATOR_FACTORIES, and get the @OptIn(@ExperimentalWindowFunctions::class) annotation. 
(The latter point is a follow up to what I said before. @OptIn will be more neat on WindowRelationalOperatorFactoryDefault than on DEFAULT_RELATIONAL_OPERATOR_FACTORIES)

**Review by vgapeyev — CHANGES_REQUESTED** [review]
The @OptIn part is largely all right, but let's see how this looks without the `experimental()` method, as we discussed.

File: lang/src/test/kotlin/org/partiql/lang/eval/evaluatortestframework/PartiQLCompilerPipelineFactory.kt [hunk 2]
```diff
@@ -73,19 +74,22 @@ internal class PartiQLCompilerPipelineFactory : PipelineFactory {
             typedOpBehavior = evaluatorOptions.typedOpBehavior
         )
 
+        @OptIn(ExperimentalWindowFunctions::class)
+        val compiler = if (experimentalFeatures) PartiQLCompilerBuilder.experimental() else PartiQLCompilerBuilder.standard()
```
Comment (vgapeyev) [inline lang/src/test/kotlin/org/partiql/lang/eval/evaluatortestframework/PartiQLCompilerPipelineFactory.kt:78]: I expect no annotation will be needed here after `experimental()` method is gone.

**Review by vgapeyev — COMMENTED** [review]
A couple catch-up comments.

File: lang/src/main/kotlin/org/partiql/annotations/PartiQLExperimental.kt [hunk 3]
```diff
@@ -8,5 +8,6 @@ package org.partiql.annotations
  * To read more on the opt-in requirements, see https://kotlinlang.org/docs/opt-in-requirements.html#opt-in-to-using-api.
  */
 
+@Deprecated("We are moving toward feature level Opt-in requirement. This opt-in requirement is not used anymore. Remove its usages from your code.")
 @RequiresOptIn(message = "This feature is experimental. It may be changed in the future without notice.", level = RequiresOptIn.Level.ERROR)
 annotation class PartiQLExperimental
```
[the remark this answers is not in the document]
Comment (yliuuuu) [reply lang/src/main/kotlin/org/partiql/annotations/PartiQLExperimental.kt]: Leave it in for now. I don't have a strong opinion on this.

File: lang/src/test/kotlin/org/partiql/lang/compiler/memorydb/QueryEngine.kt [hunk 4]
```diff
@@ -34,7 +34,8 @@ internal const val DB_CONTEXT_VAR = "in-memory-database"
 /**
  * This class is a demonstration of how to integrate a storage layer with the experimental PartiQLCompilerPipeline.
  */
-@PartiQLExperimental
+
+@OptIn(ExperimentalPartiQLCompilerPipeline::class)
```
[the remark this answers is not in the document]
Comment (yliuuuu) [reply lang/src/test/kotlin/org/partiql/lang/compiler/memorydb/QueryEngine.kt:37]: I would argue for an internal test class like this, applying annotation for class level makes maintenance a bit easier.

File: lang/src/main/kotlin/org/partiql/lang/eval/physical/PhysicalBexprToThunkConverter.kt [hunk 5]
```diff
@@ -295,7 +295,7 @@ internal class PhysicalBexprToThunkConverter(
     }
 
     // TODO: Remove from experimental once https://github.com/partiql/partiql-docs/issues/31 is resolved and a RFC is approved
-    @PartiQLExperimental
+    @ExperimentalWindowFunctions
     override fun convertWindow(node: PartiqlPhysical.Bexpr.Window): RelationThunkEnv {
```
[the remark this answers is not in the document]
Comment (yliuuuu) [reply lang/src/main/kotlin/org/partiql/lang/eval/physical/PhysicalBexprToThunkConverter.kt:298]: I believe this should be an `optin`.

**Review by vgapeyev — APPROVED** [review]
Looks good, but see some follow-up comments.
Also, please update the PR description to reflect the final state.  I hope the informative part of it can make its way into the commit message of the merge.

**Pull request merged**